### ISSUE TITLE
Join convex hulls

### ISSUE THREAD

**Issue by skarnproject** [kmammou/v-hacd#53]
Is it somehow feasible to join the resulted set of convex hull into one solid object without intersecting geometry? I am more than satisfied with outer shape that this addon does, however, there is bunch of separate objects in the resulting output that cannot be actually merged together due to intersecting faces. That feature can be used by many people who are doing any sort of gamedev production for getting object collision shapes.

**Comment by jratcliff63367** [comment]
I'm not certain why this is something you would want?  The entire purpose of the convex decomposition process is to produce a set of compound geometries appropriate for real-time collision detection.  This is how it works in UE4 when you create collision volumes using V-HACD.

If all you want is a simplified mesh, maybe a mesh simplification tool is more what you are looking for?

**Comment by skarnproject** [comment]
> I'm not certain why this is something you would want? The entire purpose of the convex decomposition process is to produce a set of compound geometries appropriate for real-time collision detection. This is how it works in UE4 when you create collision volumes using V-HACD.
> 
> If all you want is a simplified mesh, maybe a mesh simplification tool is more what you are looking for?

The problem with mesh simplification is that it does not gurantee that the shape of the result is gonna be convex and will be able to wrap the actual geometry. V-HACD produces the perfect shapes I need for collision, the only problem is that they contain some geometry inside which add extra overhead on my collision system. I am not working in UE, but in a custom engine, and the collision detection system is very different here it seems.

**Comment by kmammou** [comment]
you can play with minVolumePerCHas described here http://kmamou.blogspot.com/2014/12/v-hacd-adaptive-convex-hulls-sampling.html

**Comment by skarnproject** [comment]
It seems you got me wrong. I am absolutely fine with the number of vertices in the resulted hull. The problem is that the resulted hull is basically a bunch of separate convex shapes. I need to find a way to run a sorto of boolean operation to make these pieces of geometry merge into one without any faces inside. Would that be possible to do somehow?

**Comment by tobyndax** [comment]
I think you, in detail, need to explain what you plan to use the end result for, and why the initial mesh you are giving to V-HACD, is not good enough for that purpose. 
As mentioned by @jratcliff63367 the point of decomposing objects is to get multiple convex shapes so that you can use them for physics simulations. Most physics simulations are quite bad with concave shapes as collision shapes, but good with convex shapes and can often simulate multiple rigidly joint convex shapes. 
At least I have a hard time understanding _why_ you would want to rejoin the meshes as this would once again produce a concave mesh and thus usually not improve the physics simulation. 

p.s. 
Your input mesh (concave, single) -> Result from V-HACD (convex, multiple) -> Boolean join remove internal faces (concave, single)

**Comment by skarnproject** [comment]
> I think you, in detail, need to explain what you plan to use the end result for, and why the initial mesh you are giving to V-HACD, is not good enough for that purpose.
> As mentioned by @jratcliff63367 the point of decomposing objects is to get multiple convex shapes so that you can use them for physics simulations. Most physics simulations are quite bad with concave shapes as collision shapes, but good with convex shapes and can often simulate multiple rigidly joint convex shapes.
> At least I have a hard time understanding _why_ you would want to rejoin the meshes as this would once again produce a concave mesh and thus usually not improve the physics simulation.
> 
> p.s.
> Your input mesh (concave, single) -> Result from V-HACD (convex, multiple) -> Boolean join remove internal faces (concave, single)

The engine I work with does collision detection through some BSP Tree algorithm, I don't know how it exactly works. But when I use these shapes it becomes laggy. The joined result of what V-HACD does basically produces a simplifed mesh that wraps the entire original mesh, yet ignores small details. The mesh simplification and decimation algorithms that I know do not preserve the original shape which is not something that I need.

**Comment by tobyndax** [comment]
I would suggest you test this in something like Blender. I am still quite unsure that this will help your performance in the end. 
(https://blender.stackexchange.com/questions/71687/removing-internal-geometry) 

Also, am I right to assume that when you say preserve original shape you mean that it encapsulate (i.e. all original data-points lie withing the new hull), as opposed to preserve as in being the exact same, which V-HACD and mesh-simplification methods in general doesn't do. 

For the Physics Engines I've used in the past the experience has been something like the following:
1. Using the full mesh for collision -> Horrible Performance
2. Using a convex hull for collision -> Good performance but too inaccurate collisions
3. Decompose to multiple convex hulls. Create a compound shape, add each child object disable collision between object within the same  shape. -> Decent performance, quite accurate collisions.

**Comment by skarnproject** [comment]
Yeah, that would be accurate for physics engines system. Our collision works through binary space partitioning. It is not horrible with full mesh for collision, but sinplified geometry is prefered. Intersecting geometry in collision leads to very long export process (tons of recursive calculations) and _worse_ performance in the game.

I am sure about what I need, since we already have assets with that collison but do not currently have access to software that was able to produce them. 

The boolean operations in Blender do not work, because geoemetry does not intersect precisely.
